bazaarvoice/emodb
# Streaming improvements for splits and scans

## Github Issue #

None

## What Are We Doing Here?

There are two API for streaming table records from Emo: scan and getSplit.  Both of these filter out deleted documents to the caller; that is, documents whose most recent delta indicates a deletion.  When deletes are sporadic this isn't a problem.  However, consider the extreme case of a table with 10M documents, all of which are deleted by their most recent delta.  When an API call is made to Emo to get a scan or split for this table Emo will traverse all 10M documents in the case of scan, or the full split size in the case of a split, iterating to the first undeleted document.  Note that the user provided "limit" parameter doesn't matter; even with a limit of 1 the entire table will be scanned before determining there are no results.  This scan can take minutes and more, causing client timeouts.

This PR makes two changes:

1. It allows the caller to explicitly set as part of the API whether a scan or split should include deleted documents.  This is false by default for backwards compatibility.
2. When returning a scan or split with many deleted documents Emo will inject whitespace into the JSON array stream intermittently to keep the connection alive.  Because the whitespace is inserted between documents it should still be readable by any compliant JSON parser.

As a bonus this same process is also being used when getting a list of tables from the API.  This way if a user only has permission to read a small subset of tables the response won't block until the first readable table is scanned.

## How to Test and Verify

1. Check out this PR
2. Create a table and write multiple records to the table.
3. Delete several records
3. To a scan and split on the table, both with and without the "includeDeletes" parameter, and verify the results are consistent.

## Risk

This is a fairly low-risk change.  The biggest risk is if some JSON parser implementation chokes on the whitespace (which it shouldn't).  It's possible with buffering the whitespace solution may not work it all cases, but if it doesn't then the outcome is unchanged from present.

### Level 

`Medium`

### Required Testing

`Manual`

### Risk Summary

Add one or a few complete sentences about the possible risks or concerns for
this change.

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
